### DISCUSSION TITLE
undo and redo in jointJs

### DISCUSSION THREAD

**Discussion by Mohanakrishnan96** [clientIO/joint#1654] [Q&A] [answered]
Hi,
I am trying to implement undo and redo functionality in jointJs, but couldnt find anything related to that.
is this functionality available in jointJs or we need to use jointJs+

**Comment by kumilingus** [comment, marked as answer]
The undo/redo functionality is available only in `JointJS+` (https://resources.jointjs.com/docs/rappid/v3.5/dia.CommandManager.html).

The `CommandManager` plugin from `JointJS+` listens for all changes to the graph and saves the deltas of the changes. Here's a simplified example (the plugin also handles cell addition/removal, recognizes batch changes and filters redundant changes).
```js
const changes = [];
graph.on('change', cell => changes.push([cell.id, cell.changed]));

function undo() {
    if (changes.length === 0) return;
    const [id, change] = changes.pop();
    graph.getCell(id).set(change);
}
```
